libvast/src/system/index.cpp

```diff
@@ -1253,66 +1253,81 @@ index(index_actor::stateful_pointer<index_state> self,
-                          atom::replace_v, old_partition_id, new_partition_id,
-                          aps.synopsis)
-                .then(
-                  [self, rp, old_partition_id, new_partition_id,
-                   result](atom::ok) mutable {
-                    self->state.persisted_partitions.insert(new_partition_id);
-                    self
-                      ->request(static_cast<index_actor>(self), caf::infinite,
-                                atom::erase_v, old_partition_id)
-                      .then(
-                        [=](atom::done) mutable {
-                          rp.deliver(result);
-                        },
-                        [=](const caf::error& e) mutable {
-                          rp.deliver(e);
+              // other ones are "deleted" from the meta index. If the new
+              // partition is empty, all partitions are deleted.
+              auto start = size_t{0};
+              if (aps.synopsis) {
+                start = 1;
+                auto old_partition_id = old_partition_ids.at(0);
+                self
+                  ->request(self->state.meta_index, caf::infinite,
+                            atom::replace_v, old_partition_id, new_partition_id,
+                            aps.synopsis)
+                  .then(
+                    [self, rp, old_partition_id, new_partition_id,
+                     result](atom::ok) mutable {
+                      self->state.persisted_partitions.insert(new_partition_id);
+                      self
+                        ->request(static_cast<index_actor>(self), caf::infinite,
+                                  atom::erase_v, old_partition_id)
+                        .then(
+                          [=](atom::done) mutable {
+                            rp.deliver(result);
+                          },
+                          [=](const caf::error& e) mutable {
+                            rp.deliver(e);
+                          });
+                    },
+                    [rp](const caf::error& e) mutable {
+                      rp.deliver(e);
+                    });
+              } else {
+                rp.deliver(result);
+              }
+              for (size_t i = start; i < old_partition_ids.size(); ++i) {
+                auto partition_id = old_partition_ids[i];
+                self
+                  ->request(self->state.meta_index, caf::infinite,
+                            atom::erase_v, partition_id)
+                  .then([](atom::ok) { /* nop */ },
+                        [](const caf::error& e) {
+                          VAST_WARN("index failed to erase {} from meta index",
+                                    e);
                         });
-                  },
-                  [rp](const caf::error& e) mutable {
-                    rp.deliver(e);
-                  });
-            }
-            for (size_t i = 1; i < old_partition_ids.size(); ++i) {
```

Comment:
Btw, it looked to me like a separate bug that the `for`-loop was outside the scope of the `else`-branch so i moved it inside, but please verify @6yozo / @dominiklohmann .

Reply:
Yes, almost certainly a bug. Thanks for noticing/fixing. Hopefully I can find the time to finish the co_result<T> stuff on the sooner side such that we can avoid similar issues caused by the nesting depth making the code harder to follow in the future.

Feel free to resolve.